File: app/strategies/rsi_strategy.py
from collections import defaultdict

import pandas as pd

from app.signals.signal_publisher import (
    SignalPublisher
)

from app.core.logger import logger


class RSIStrategy:

    def __init__(
        self,
        period=14
    ):

        self.period = period

        self.price_history = defaultdict(list)

    async def generate_signal(
        self,
        symbol,
        price
    ):

        history = self.price_history[symbol]

        history.append(price)

        if len(history) < self.period:
            return

        prices = pd.Series(history)

        delta = prices.diff()

        gain = (
            delta.where(delta > 0, 0)
            .rolling(self.period)
            .mean()
        )

        loss = (
            -delta.where(delta < 0, 0)
            .rolling(self.period)
            .mean()
        )

        rs = gain / loss

        rsi = (
            100
            - (
                100 / (1 + rs)
            )
        )

        latest_rsi = rsi.iloc[-1]

        signal = None

        if latest_rsi < 30:
            signal = "BUY"

        elif latest_rsi > 70:
            signal = "SELL"

        if signal:

            payload = {
                "strategy": "rsi",
                "symbol": symbol,
                "signal": signal,
                "price": price
            }

            logger.info(
                f"RSI Signal: {payload}"
            )

            await SignalPublisher.publish(
                payload
            )

File: backend/src/trading/strategies/rsiStrategy.ts
import { BaseStrategy } from "./baseStrategy";
import { SignalPublisher } from "../signals/signalPublisher";
import { calculateRSI } from "../../utils/indicators";
import { logger } from "../../core/logger";

export class RSIStrategy extends BaseStrategy {
  private period: number;
  private priceHistory: { [symbol: string]: number[] } = {};

  constructor(config?: { period?: number }) {
    super();
    this.period = config?.period || 14;
  }

  async generateSignal(symbol: string, price: number): Promise<void> {
    if (!this.priceHistory[symbol]) {
      this.priceHistory[symbol] = [];
    }

    const history = this.priceHistory[symbol];
    history.push(price);

    // Keep history clean and capped to avoid memory leaks
    const maxHistory = this.period + 100;
    if (history.length > maxHistory) {
      history.shift();
    }

    if (history.length < this.period) {
      return;
    }

    const rsi = calculateRSI(history, this.period);
    const latestRsi = rsi[rsi.length - 1];

    let signal: string | null = null;

    if (latestRsi < 30) {
      signal = "BUY";
    } else if (latestRsi > 70) {
      signal = "SELL";
    }

    if (signal) {
      const payload = {
        strategy: "rsi",
        symbol,
        signal,
        price,
      };

      logger.info(`RSI Signal: ${JSON.stringify(payload)}`);
      await SignalPublisher.publish(payload);
    }
  }
}

// Register strategy
BaseStrategy.register("rsi", RSIStrategy);
export default RSIStrategy;
